launch_handler-explainer.md

```diff
@@ -0,0 +1,141 @@
+# Web App Launch Handling
+
+## Overview
+
+This document describes a new `launch_handler` manifest member that enables
+web apps to customise their launch behaviour across all types of app launch
+triggers.
+
+
+## Use Cases
+
+- Web apps that are designed to be used in a single window e.g. a music app.
+
+- Web app that capture and handle share target events and user navigations
+  in existing windows without invoking a navigation and losing existing state.
+  E.g. sharing an image to a chat web app could open a picker overlayed on top
+  of an existing conversation to select which contact to send it to.
+
+
+## Background
+
+There are several ways for a web app window to be opened:
+- Platform specific app launch surface
+- [File handling](https://github.com/WICG/file-handling/blob/main/explainer.md)
+- [Share target](https://w3c.github.io/web-share-target/)
+- [In scope link capturing](https://github.com/WICG/sw-launch/blob/master/declarative_link_capturing.md)
+- [Shortcuts](https://www.w3.org/TR/appmanifest/#dfn-shortcuts)
+- [Note taking](https://wicg.github.io/manifest-incubations/index.html#note_taking-member)
+- [Protocol handling](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/URLProtocolHandler/explainer.md)
+
+Web apps launched via these triggers will open in a new or existing app window
+depending on the user agent platform. There is currently no mechanism for the
+web app to configure this behaviour.
+
+
+## Proposal
+
+- Add a `launch_handler` member to the web app manifest specifying the default
+  client selection and navigation behaviour for web app launches.
+  The shape of this member is as follows:
+  ```
+  "launch_handler": {
+    "route_to_client": null | "new" | "existing",
+    "navigate_client": null | true | false
+  }
+  ```
+
+  If unspecified then `launch_handler` defaults to
+  `{"route_to_client": null, "navigate_client": null}` whereby the behaviour
```

Comment:
wouldn't that not be undefined in JS sense then? It is JSON after all

Reply:
Do you mean undefined instead of null or undefined instead of `{"route_to_client": null, "navigate_client": null}`?

Reply:
well, if launch_handler is undefined then so are launch_handler.route_to_client and launch_handler.navigate_client

I don't see any reason to introduce null here.

Reply:
It makes more sense with this part of the proposal:
https://github.com/alancutter/manifest-incubations/blob/launch-handler/launch_handler-explainer.md#future-extensions-to-this-proposal
```
{
  "name": "Example app",
  "description": "This app will navigate existing clients unless it was launched via the share target API.",
  "launch_handler": {
    "route_to_client": "existing",
    "navigate_client": true
  },
  "share_target": {
    "action": "share.html",
    "params": {
      "title": "name",
      "text": "description",
      "url": "link"
    },
    "launch_handler": {
      "navigate_client": false
    }
  }
}
```

`manifest.share_target.launch_handler.route_to` is null meaning to use the value of `manifest.launch_handler.route_to`.

Reply:
It could be like (pseudocode):
```
interface Manifest {
  launch_handler: LaunchHandler,
  share_target: ShareTarget,
  ...
}

interface LaunchHandler {
  route_to: "auto" | "new-client" | "existing-client",
  navigate_client: "auto" | Boolean,
}

interface ShareTarget {
  launch_handler: LaunchHandlerDelta,
  ...
}

interface LaunchHandlerDelta = LaunchHandler except all the fields can also be null.
```

Reply:
Just a point of clarification: I believe @kenchris is merely saying that these should be `undefined` rather than `null` -- which are two distinct values in JS with different meaning. It makes sense to say that `launch_handler: undefined` is equivalent to `launch_handler: {x: undefined, y: undefined}`, not so much to have it imply `null`.